[MAHOUT-1896] Add convenience methods for interacting with Spark ML

rawkintrevo opened the pull request:
Currently the method for ingesting RDDs to DRM is `drmWrap`. This is a flexible method, however there are many cases when the RDD to be wrapped is either RDD[org.apache.spark.mllib.lingalg.Vector], RDD[org.apache.spark.mllib.regression.LabeledPoint], or DataFrame[Row] (as is the case when working with SparkML. It makes sense to create convenience methods for converting these types to DRM.

* [x] Add `drmWrapMLLibLabeledPoint`
* [x] Add `drmWrapDataFrame`
* [x] Add `drmWrapMLLibVector`
* [x] Tests


Minor improvement

[Relevent JIRA MAHOUT-1896](https://issues.apache.org/jira/browse/MAHOUT-1896)

Requires no license updates

andrewpalumbo commented on spark/src/main/scala/org/apache/mahout/sparkbindings/package.scala:
```diff
@@ -141,6 +144,42 @@ package object sparkbindings {
+                   cacheHint: CacheHint.CacheHint = CacheHint.NONE,
+                   canHaveMissingRows: Boolean = false): CheckpointedDrm[Int] = {
+    val drmRDD: DrmRdd[Int] = rdd.zipWithIndex.map( lv => (lv._2.toInt,
+      new org.apache.mahout.math.DenseVector( Array(lv._1.label) ++ lv._1.features.toArray)) )
```
I don't think that this should always be dense

andrewpalumbo replied:
I think that the label should be appended to the end rather than the beginning.

andrewpalumbo commented on spark/src/test/scala/org/apache/mahout/sparkbindings/drm/DrmLikeSuite.scala:
```diff
@@ -63,6 +65,65 @@ class DrmLikeSuite extends FunSuite with DistributedSparkSuite with DrmLikeSuite
+    import org.apache.spark.mllib.linalg.{Vectors => SparkVector}
+    import org.apache.spark.mllib.regression.LabeledPoint
+
+    val sc = mahoutCtx.asInstanceOf[SparkDistributedContext].sc
```
A `MahoutDistributedContext` should be implicitly converted into a SparkDistributedContext when it is needed.  This could be written as : 

     ``` val sc:SparkDistributedContext```

andrewpalumbo commented on spark/src/test/scala/org/apache/mahout/sparkbindings/drm/DrmLikeSuite.scala:
```diff
@@ -63,6 +65,65 @@ class DrmLikeSuite extends FunSuite with DistributedSparkSuite with DrmLikeSuite
+                                   LabeledPoint(2.0, SparkVector.dense(3.0, 0.0, 5.0)),
+                                   LabeledPoint(3.0, SparkVector.dense(4.0, 0.0, 6.0)) ))
+
+    val lpDRM = drmWrapMLLibLabeledPoint(rdd= lpRDD)
```
``` rdd = lmRDD``` (missing space before equals)

andrewpalumbo commented on spark/src/test/scala/org/apache/mahout/sparkbindings/drm/DrmLikeSuite.scala:
```diff
@@ -63,6 +65,65 @@ class DrmLikeSuite extends FunSuite with DistributedSparkSuite with DrmLikeSuite
+
+    import org.apache.spark.mllib.linalg.{Vectors => SparkVector}
+
+    val sc = mahoutCtx.asInstanceOf[SparkDistributedContext].sc
```
same implicit conversion as before.

andrewpalumbo commented on spark/src/main/scala/org/apache/mahout/sparkbindings/package.scala:
```diff
@@ -141,6 +144,42 @@ package object sparkbindings {
+    val drmRDD: DrmRdd[Int] = df.rdd
+                                .zipWithIndex
+                                .map( o => (o._2.toInt, o._1.mkString(",").split(",").map(s => s.toDouble)) )
+                                .map(o => (o._1,new org.apache.mahout.math.DenseVector( o._2 )))
```
space after `0._1,`

andrewpalumbo commented:
+1- made a few style comments and a couple about the Implicit Conversion of a MahoutDistributedContext to a SparkContext.

rawkintrevo commented:
@andrewpalumbo thanks for the review!

Fixed typos per your request and updated to convert to sparse vectors in cases where original vector was sparse. 

re: the implicit conversion of the mahoutCtx, that threw errors so I left it as was.

andrewpalumbo commented:
LGTM, +1.

rawkintrevo commented:
Will squash
